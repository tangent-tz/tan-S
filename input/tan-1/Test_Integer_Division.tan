main {
    const x := 10;
    const y := 5;

    # Test 1: division by zero using const variables
#    const a := 10;
#    const b := 0;
#    print a / b;  # Expected output: Division by zero error
#    print \n;

    # Test 2: division resulting in an integer using const variables
    const c := 15;
    const d := 4;
    print c / d;  # Expected output: 3
    print \n;

    # Test 3: division resulting in a float using const variables
    const e := 7;
    const f := 2;
    print e / f;  # Expected output: 3.5
    print \n;

    # Test 4: division with negative numbers using const variables
    const g := -20;
    const h := 5;
    print g / h;  # Expected output: -4
    print \n;

    # Test 5: division with mixed positive and negative numbers using const variables
    const i := 25;
    const j := -4;
    print i / j;  # Expected output: -6
    print \n;

    # Test 6: division by zero using var variables
#    var vara := 10;
#    var varb := 0;
#    print vara / varb;  # Expected output: Division by zero error
#    print \n;

    # Test 7: division resulting in an integer using var variables
    var varc := 15;
    var vard := 4;
    print varc / vard;  # Expected output: 3
    print \n;

    # Test 8: division resulting in a float using var variables
    var vare := 7;
    var varf := 2;
    print vare / varf;  # Expected output: 3.5
    print \n;

    # Test 9: division with negative numbers using var variables
    var varg := -20;
    var varh := 5;
    print varg / varh;  # Expected output: -4
    print \n;

    # Test 10: division with mixed positive and negative numbers using var variables
    var vari := 25;
    var varj := -4;
    print vari / varj;  # Expected output: -6
    print \n;
}
