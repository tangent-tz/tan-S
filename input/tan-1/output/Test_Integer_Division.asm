        Jump         $$main                    
        DLabel       $eat-location-zero        
        DataZ        8                         
        DLabel       $print-format-integer     
        DataC        37                        %% "%d"
        DataC        100                       
        DataC        0                         
        DLabel       $print-format-boolean     
        DataC        37                        %% "%s"
        DataC        115                       
        DataC        0                         
        DLabel       $print-format-float       
        DataC        37                        %% "%f"
        DataC        102                       
        DataC        0                         
        DLabel       $print-format-character   
        DataC        37                        %% "%c"
        DataC        99                        
        DataC        0                         
        DLabel       $print-format-string      
        DataC        37                        %% "%s"
        DataC        115                       
        DataC        0                         
        DLabel       $print-format-newline     
        DataC        10                        %% "\n"
        DataC        0                         
        DLabel       $print-format-space       
        DataC        32                        %% " "
        DataC        0                         
        DLabel       $print-format-tab         
        DataC        9                         %% "\t"
        DataC        0                         
        DLabel       $boolean-true-string      
        DataC        116                       %% "true"
        DataC        114                       
        DataC        117                       
        DataC        101                       
        DataC        0                         
        DLabel       $boolean-false-string     
        DataC        102                       %% "false"
        DataC        97                        
        DataC        108                       
        DataC        115                       
        DataC        101                       
        DataC        0                         
        DLabel       $errors-general-message   
        DataC        82                        %% "Runtime error: %s\n"
        DataC        117                       
        DataC        110                       
        DataC        116                       
        DataC        105                       
        DataC        109                       
        DataC        101                       
        DataC        32                        
        DataC        101                       
        DataC        114                       
        DataC        114                       
        DataC        111                       
        DataC        114                       
        DataC        58                        
        DataC        32                        
        DataC        37                        
        DataC        115                       
        DataC        10                        
        DataC        0                         
        Label        $$general-runtime-error   
        PushD        $errors-general-message   
        Printf                                 
        Halt                                   
        DLabel       $errors-int-divide-by-zero 
        DataC        105                       %% "integer divide by zero"
        DataC        110                       
        DataC        116                       
        DataC        101                       
        DataC        103                       
        DataC        101                       
        DataC        114                       
        DataC        32                        
        DataC        100                       
        DataC        105                       
        DataC        118                       
        DataC        105                       
        DataC        100                       
        DataC        101                       
        DataC        32                        
        DataC        98                        
        DataC        121                       
        DataC        32                        
        DataC        122                       
        DataC        101                       
        DataC        114                       
        DataC        111                       
        DataC        0                         
        Label        $$i-divide-by-zero        
        PushD        $errors-int-divide-by-zero 
        Jump         $$general-runtime-error   
        DLabel       $errors-float-divide-by-zero 
        DataC        102                       %% "float divide by zero"
        DataC        108                       
        DataC        111                       
        DataC        97                        
        DataC        116                       
        DataC        32                        
        DataC        100                       
        DataC        105                       
        DataC        118                       
        DataC        105                       
        DataC        100                       
        DataC        101                       
        DataC        32                        
        DataC        98                        
        DataC        121                       
        DataC        32                        
        DataC        122                       
        DataC        101                       
        DataC        114                       
        DataC        111                       
        DataC        0                         
        Label        $$f-divide-by-zero        
        PushD        $errors-float-divide-by-zero 
        Jump         $$general-runtime-error   
        DLabel       $usable-memory-start      
        DLabel       $global-memory-block      
        DataZ        72                        
        Label        $$main                    
        PushD        $global-memory-block      
        PushI        0                         
        Add                                    %% x
        PushI        10                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        4                         
        Add                                    %% y
        PushI        5                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        8                         
        Add                                    %% c
        PushI        15                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        12                        
        Add                                    %% d
        PushI        4                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        8                         
        Add                                    %% c
        LoadI                                  
        PushD        $global-memory-block      
        PushI        12                        
        Add                                    %% d
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-1-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        16                        
        Add                                    %% e
        PushI        7                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        20                        
        Add                                    %% f
        PushI        2                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        16                        
        Add                                    %% e
        LoadI                                  
        PushD        $global-memory-block      
        PushI        20                        
        Add                                    %% f
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-2-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        24                        
        Add                                    %% g
        PushI        20                        
        Negate                                 
        StoreI                                 
        PushD        $global-memory-block      
        PushI        28                        
        Add                                    %% h
        PushI        5                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        24                        
        Add                                    %% g
        LoadI                                  
        PushD        $global-memory-block      
        PushI        28                        
        Add                                    %% h
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-3-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        32                        
        Add                                    %% i
        PushI        25                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        36                        
        Add                                    %% j
        PushI        4                         
        Negate                                 
        StoreI                                 
        PushD        $global-memory-block      
        PushI        32                        
        Add                                    %% i
        LoadI                                  
        PushD        $global-memory-block      
        PushI        36                        
        Add                                    %% j
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-4-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        40                        
        Add                                    %% varc
        PushI        15                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        44                        
        Add                                    %% vard
        PushI        4                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        40                        
        Add                                    %% varc
        LoadI                                  
        PushD        $global-memory-block      
        PushI        44                        
        Add                                    %% vard
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-5-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        48                        
        Add                                    %% vare
        PushI        7                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        52                        
        Add                                    %% varf
        PushI        2                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        48                        
        Add                                    %% vare
        LoadI                                  
        PushD        $global-memory-block      
        PushI        52                        
        Add                                    %% varf
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-6-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        56                        
        Add                                    %% varg
        PushI        20                        
        Negate                                 
        StoreI                                 
        PushD        $global-memory-block      
        PushI        60                        
        Add                                    %% varh
        PushI        5                         
        StoreI                                 
        PushD        $global-memory-block      
        PushI        56                        
        Add                                    %% varg
        LoadI                                  
        PushD        $global-memory-block      
        PushI        60                        
        Add                                    %% varh
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-7-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        PushD        $global-memory-block      
        PushI        64                        
        Add                                    %% vari
        PushI        25                        
        StoreI                                 
        PushD        $global-memory-block      
        PushI        68                        
        Add                                    %% varj
        PushI        4                         
        Negate                                 
        StoreI                                 
        PushD        $global-memory-block      
        PushI        64                        
        Add                                    %% vari
        LoadI                                  
        PushD        $global-memory-block      
        PushI        68                        
        Add                                    %% varj
        LoadI                                  
        Duplicate                              
        JumpFalse    $$i-divide-by-zero        
        Label        -divide-8-notZero         
        Divide                                 
        PushD        $print-format-integer     
        Printf                                 
        PushD        $print-format-newline     
        Printf                                 
        Halt                                   
