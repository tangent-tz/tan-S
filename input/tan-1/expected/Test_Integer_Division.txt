0.5
5
2.5
-3.5
Division by zero error
333
